OpenRA.Mods.Common/Traits/AutoCarryall.cs

```diff
@@ -96,11 +96,35 @@ void FindCarryableForTransport(Actor self)
 				if (IsBestAutoCarryallForCargo(self, p.Actor) && ReserveCarryable(self, p.Actor))
 				{
 					busy = true;
-					self.QueueActivity(false, new PickupUnit(self, p.Actor, 0));
-					self.QueueActivity(true, new DeliverUnit(self, Target.FromCell(self.World, p.Trait.Destination.Value), Info.DropRange));
+					self.QueueActivity(false, new FerryUnit(self, p.Actor));
 					break;
 				}
 			}
 		}
+
+		class FerryUnit : Activity
+		{
+			readonly Actor cargo;
+
+			public FerryUnit(Actor self, Actor cargo)
+			{
+				this.cargo = cargo;
+			}
+
+			protected override void OnFirstRun(Actor self)
+			{
+				QueueChild(new PickupUnit(self, cargo, 0));
+			}
+
+			public override bool Tick(Actor self)
+			{
+				var dropRange = self.Trait<Carryall>().Info.DropRange;
+				var destination = cargo.Trait<Carryable>().Destination;
```

Comment:
Wouldn't it make sense to store these in in the constructor?

Reply:
It doesn't make a difference performance-wise because this will only ever tick once.